File: s29_natural_language_processing.py
# Natural Language Processing

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset [§29 Lect191: "NLP in Python - pt1"]
dataset = pd.read_csv('data/s29_Restaurant_Reviews.tsv', delimiter = '\t', quoting = 3)

# Cleaning the texts [§29 Lect192: "NLP in Python - pt2"] & [§29 Lect193: "NLP in Python - pt3"]
# & [§29 Lect194: "NLP in Python - pt4"] & [§29 Lect195: "NLP in Python - pt5"]
# & [§29 Lect196: "NLP in Python - pt6"] & [§29 Lect197: "NLP in Python - pt7"]
import re                                                    # "NLP in Python - pt2"
import nltk                                                  # "NLP in Python - pt4"
nltk.download('stopwords')                                   # "NLP in Python - pt4"
from nltk.corpus import stopwords                            # "NLP in Python - pt4"
from nltk.stem.porter import PorterStemmer                   # "NLP in Python - pt5"
corpus = []                                                  # "NLP in Python - pt7"
for i in range(0, 1000):                                     # "NLP in Python - pt7"
    review = re.sub('[^a-zA-Z]', ' ', dataset['Review'][i])  # "NLP in Python - pt2" & 'pt7'
    review = review.lower()                                  # "NLP in Python - pt3"
    review = review.split()                                  # "NLP in Python - pt4"
    ps = PorterStemmer()                                     # "NLP in Python - pt5"
    review = [ps.stem(word) for word in review if not word in set(stopwords.words('english'))]
    review = ' '.join(review)                                # "NLP in Python - pt6"
    corpus.append(review)                                    # "NLP in Python - pt7"

# Creating the Bag of Words model [§29 Lect198: "NLP in Python - pt8"] & [§29 Lect199: "NLP in Python - pt9"]
from sklearn.feature_extraction.text import CountVectorizer
cv = CountVectorizer(max_features = 1500)
X = cv.fit_transform(corpus).toarray()
y = dataset.iloc[:, 1].values

# Splitting the dataset into the Training set and Test set [§29 Lect200: "NLP in Python - pt10"]
from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, random_state = 0)

# Fitting Naive Bayes to the Training set [§29 Lect200: "NLP in Python - pt10"]
from sklearn.naive_bayes import GaussianNB
classifier = GaussianNB()
classifier.fit(X_train, y_train)

# Predicting the Test set results [§29 Lect200: "NLP in Python - pt10"]
y_pred = classifier.predict(X_test)

# Making the Confusion Matrix [§29 Lect200: "NLP in Python - pt10"]
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)

File: s29_natural_language_processing.R
# Natural Language Processing

# Importing the dataset [§29 Lect202: "NLP in R - pt1"]
dataset_original = read.delim('data/s29_Restaurant_Reviews.tsv', quote = '', stringsAsFactors = FALSE)

# Cleaning the texts [§29 Lect203: "NLP in R - pt2"] & [§29 Lect204: "NLP in R - pt3"]
# & [§29 Lect205: "NLP in R - pt4"] & [§29 Lect206: "NLP in R - pt5"] & [§29 Lect207: "NLP in R - pt6"]
# & [§29 Lect208: "NLP in R - pt7"] & [§29 Lect209: "NLP in R - pt8"]
# install.packages('tm')
# install.packages('SnowballC')
library(tm)                                                   # "NLP in R - pt2"
library(SnowballC)                                            # "NLP in R - pt6" ...Unnecessary?
corpus = VCorpus(VectorSource(dataset_original$Review))       # "NLP in R - pt2"
corpus = tm_map(corpus, content_transformer(tolower))         # "NLP in R - pt3"
corpus = tm_map(corpus, removeNumbers)                        # "NLP in R - pt4"
corpus = tm_map(corpus, removePunctuation)                    # "NLP in R - pt5"
corpus = tm_map(corpus, removeWords, stopwords())             # "NLP in R - pt6"
corpus = tm_map(corpus, stemDocument)                         # "NLP in R - pt7"
corpus = tm_map(corpus, stripWhitespace)                      # "NLP in R - pt8"

# Creating the Bag of Words model [§29 Lect210: "NLP in R - pt9"]
dtm = DocumentTermMatrix(corpus)
dtm = removeSparseTerms(dtm, 0.999)
dataset = as.data.frame(as.matrix(dtm))                       # "NLP in R - pt10"
dataset$Liked = dataset_original$Liked                        # "NLP in R - pt10"


# Encoding the target feature as factor                       # "NLP in R - pt10"
dataset$Liked = factor(dataset$Liked, levels = c(0, 1))

# Splitting the dataset into the Training set and Test set    # "NLP in R - pt10"
# install.packages('caTools')
library(caTools)
set.seed(123)
split = sample.split(dataset$Liked, SplitRatio = 0.8)
training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)

# Fitting Random Forest Classification to the Training set    # "NLP in R - pt10"
# install.packages('randomForest')
library(randomForest)
classifier = randomForest(x = training_set[-692],
                          y = training_set$Liked,
                          ntree = 10)

# Predicting the Test set results                             # "NLP in R - pt10"
y_pred = predict(classifier, newdata = test_set[-692])

# Making the Confusion Matrix                                 # "NLP in R - pt10"
cm = table(test_set[, 692], y_pred)